connector_salesforce/unit/binder.py

```diff
@@ -0,0 +1,94 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    Author: Nicolas Bessi
+#    Copyright 2014 Camptocamp SA
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+from __future__ import absolute_import
+from openerp.osv import fields
+from openerp.addons.connector.connector import Binder
+from ..backend import salesforce_backend
+
+
+@salesforce_backend
+class SalesforeceBinder(Binder):
+    """ Manage bindings between Models identifier and Salesforce identifier"""
+    _model_name = []
+
+    def to_openerp(self, salesforce_id, unwrap=False):
+        """Returns the Odoo id for an external ID""
+
+        :param salesforce_id: salesforce_id row unique idenifier
+        :type salesforce_id: str
+
+        :param unwrap: If True returns the id of the record related
+                       to the binding record
+
+        :return: id of binding or if unwrapped the id of the record related
+                 to the binding record
+        :rtype: int
+        """
+        with self.session.change_context({'active_test': False}):
+            binding_ids = self.session.search(
+                self.model._name,
+                [('salesforce_id', '=', salesforce_id),
+                 ('backend_id', '=', self.backend_record.id)],
+
+            )
+        if not binding_ids:
+            return None
+        assert len(binding_ids) == 1, "Several records found: %s" % binding_ids
+        binding_id = binding_ids[0]
+        if unwrap:
+            return self.session.read(self.model._name,
+                                     binding_id,
+                                     ['openerp_id'])['openerp_id'][0]
+        else:
+            return binding_id
+
+    def to_backend(self, binding_id):
+        """Return the external code for a given binding model id
+
+        :param binding_id: id of a binding model
+        :type binding_id: int
+
+        :return: external code of `binding_id`
+        """
+        sf_record = self.session.read(self.model._name,
+                                      binding_id,
+                                      ['salesforce_id'])
+        assert sf_record, 'No corresponding binding found'
```

Comment:
IMO the binding should never raise an error when it does not find something so the methods can be used to know if we know the external ID, or if we know the odoo ID for an External ID when they returns None.

At the very least, if really you want to raise something, do not use an assert because having no external id in a binding is something that may and that will happen. 

See the one in the Magento connector https://github.com/OCA/connector-magento/blob/7.0/magentoerpconnect/unit/binder.py#L84